### PR TITLE
Added support for multiple exploded parameters for both server and client

### PR DESCRIPTION
**Pull request by Uniqen**
Some minor changes to UriTemplate (and tests) in order to add support for multiple exploded parameters for both server and client.

Multiple parameters is useful e.g. when forwarding filter parameters and Pageable in an API-gateway. 

Example of usage:
```java
    @Controller('/itunes')
    static class ItunesController {
        @Get("/search-exploded/multipleExplodedPojos{?term*}{&song*}")
        SearchResult searchMultipleExplodedPojos(@QueryValue SearchParams term, @QueryValue SearchParams2 song) {
...
        }

        @Get("/search-exploded/multipleOptionalExplodedPojos{?term*}{?song*}")
        SearchResult searchMultipleOptionalExplodedPojos(@QueryValue SearchParams term, @QueryValue SearchParams2 song) {
...
        }
}

 @Client("/itunes")
    static interface ItunesClient {

        @Get("/search-exploded/multipleExplodedPojos{?term*}{&song*}")
        SearchResult searchExplodedMultiplePojos(SearchParams term, SearchParams2 song);

        @Get("/search-exploded/multipleOptionalExplodedPojos{?term*}{?song*}")
        SearchResult searchExplodedMultipleOptionalPojos(@Nullable SearchParams term, @Nullable SearchParams2 song);
}
```

### CODE REVIEW

File: http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy [hunk 1]
```diff
@@ -33,24 +33,32 @@ class UriTemplateSpec extends Specification {
...
-        ''             | '/hello/{name}'      | [name: 'Fred']                          | '/hello/Fred'
-        '/test/'       | '/hello/{name}'      | [name: 'Fred']                          | '/test/hello/Fred'
-        '{var}'        | '{var2}'             | [var: 'foo', var2: 'bar']               | 'foo/bar'
-        '/book{/id}'   | '/author{/authorId}' | [id: 'foo', authorId: 'bar']            | '/book/foo/author/bar'
-        '{var}/'       | '{var2}'             | [var: 'foo', var2: 'bar']               | 'foo/bar'
-        '{var}'        | '/{var2}'            | [var: 'foo', var2: 'bar']               | 'foo/bar'
-        '{var}{?q}'    | '/{var2}'            | [var: 'foo', var2: 'bar', q: 'test']    | 'foo/bar?q=test'
-        '{var}{?q}'    | '{var2}'             | [var: 'foo', var2: 'bar', q: 'test']    | 'foo/bar?q=test'
-        '{var}{#hash}' | '{var2}'             | [var: 'foo', var2: 'bar', hash: 'test'] | 'foo/bar#test'
-        '/'            | '{?req*}'            | [req: [var: 'foo', var2: null]]         | '/?var=foo'
```
Comment (jameskleeh) [inline http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy]: Looks like this line was removed

File: http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy [hunk 2]
```diff
@@ -33,24 +33,32 @@ class UriTemplateSpec extends Specification {
...
+        '/'            | '/hello/{name}'      | [name: 'Fred']                            | '/hello/Fred'
+        ''             | '/hello/{name}'      | [name: 'Fred']                            | '/hello/Fred'
+        '/test/'       | '/hello/{name}'      | [name: 'Fred']                            | '/test/hello/Fred'
+        '{var}'        | '{var2}'             | [var: 'foo', var2: 'bar']                 | 'foo/bar'
+        '/book{/id}'   | '/author{/authorId}' | [id: 'foo', authorId: 'bar']              | '/book/foo/author/bar'
+        '{var}/'       | '{var2}'             | [var: 'foo', var2: 'bar']                 | 'foo/bar'
+        '{var}'        | '/{var2}'            | [var: 'foo', var2: 'bar']                 | 'foo/bar'
+        '{var}{?q}'    | '/{var2}'            | [var: 'foo', var2: 'bar', q: 'test']      | 'foo/bar?q=test'
+        '{var}{?q}'    | '{var2}'             | [var: 'foo', var2: 'bar', q: 'test']      | 'foo/bar?q=test'
+        '{var}{#hash}' | '{var2}'             | [var: 'foo', var2: 'bar', hash: 'test']   | 'foo/bar#test'
```
Comment (jameskleeh) [inline http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy]: Can you please revert the spacing changes here?

**Comment by jameskleeh** [comment]
Nice work! Thanks for the PR. Just some minor changes requested and we'll get this merged

File: http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy (hunk 1 above)
Comment (Uniqen) [reply http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy]: Sorry, my mistake. Good catch!

File: http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy (hunk 2 above)
Comment (Uniqen) [reply http/src/test/groovy/io/micronaut/http/uri/UriTemplateSpec.groovy]: Reverted, I had Intellij to ignore whitespaces so this one slipped through.

**Comment by Uniqen** [comment]
Thanks, and thank you for a nice promising framework.

**Comment by traycho** [comment]
Build is failing because of test `test returning a file from a controller`, I think it is broken in 1.2.x.

**Comment by Uniqen** [comment]
@jameskleeh I've made the requested changes and merged master to resolve the CI error.

**Pull request merged**